### PR TITLE
Enable Agents access to returns

### PR DESCRIPTION
**Pull request by tonymccrae-ee**
### Description of Work carried through

Adjust AuthAction to allow agents to work with PPT accounts which they have been delegated access to.

TLDR - to auth as an Agent, you need to know the client's PPT identifier BEFORE calling auth,
This is different to the existing single user code with calls auth and then finds the PPT identifier in the auth response.

Screenshot from 2022-03-18 09-29-45

## Prompt Agents for their client's PPT identifier

When auth'ing as an agent, the AuthAction might supply the PPT identifier of the account the agent is trying to access.

This has to be supplied as as a parameter to the auth call. 
This is different to the existing single user code which auths and then finds the PPT identifier from the enrolments in the auth response.

Add a new screen at `/agents` to prompt agents for the client identifier of their client.
This identifier is stashed on the MDTP session.
This screen is be protected with an agents only auth check.

If AuthAction is called and there is an agent supplied client identifier on the session then we should assume this request is from an agent and add the client identifer to the auth request.

## When to prompt for an agent supplied client id.

If a user is dropped out of AuthAction with `InsufficientEnrolments` they are redirected to a new not enroller end point at `/plastic-packaging-tax/account/not-enrolled`.

Users with insufficient enrolments are  we likely to be in one of these states:

- A single user who is not register for PPT.
They should be redirected to registration as per the existing behaviour

- An agent who has not supplied a client id
They should be redirected to `/agents` and prompted to supply one.

- An agent who has supplied an invalid client id or one which that have no delegated access too
The should be redirected back to `/agents` to supply a valid client id.

The auth client `InsufficiantEnrolments` does seem to supply any details about the authenticated users; so this new end point preforms a simple auth check with not enrolment predicate to determine who the user is.

## Reading the PPT identifier from auth enrolments does not work for agents

An agent's access to a client's enrolment is handled by a piece of auth code named `AgentDelegationPredicate`.
Importantly this code does not add the delegated enrolment to the auth enrolments response when an agent calls auth.

This means that any PPT code which is reading the PPT identifier from auth enrolments will not work with agents.
This is ok in the front end where we have access to the MTDP session and can see the user supplied client id.

#### Check list 
 - [x] `./precheck` was executed (Integration/Component/Unit tests)
 - [x] `sbt scalafmt test:scalafmt` was executed 
 - [ ] Required Environment Config has been amended/added
 - [ ] Links to dependencies have been included (BE/FE work)
 - [x] User Acceptance Tests (UAT) were run locally and have passed.
 - [ ] No Accessibility errors/warnings reported by Wave

### CODE REVIEW

File: conf/application.conf [hunk 1]
```diff
@@ -134,7 +134,7 @@ assets {
 urls {
   exitSurvey = "http://localhost:9514/feedback/plastic-packaging-tax-registration"
   login = "http://localhost:9949/auth-login-stub/gg-sign-in"
-  loginContinue = "http://localhost:8505/plastic-packaging-tax/submit-return-for-plastic-packaging-tax/submit-return"
```
Comment (tonymccrae-ee) [inline conf/application.conf:137]: This feels like the natural setting for login landing page.

Comment (tonymccrae-ee) [reply conf/application.conf:137]: This url change has been separately approved; marking as resolved.

File: test/uk/gov/hmrc/plasticpackagingtax/returns/forms/ClientIdentifierSpec.scala [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      "provided with longer than 15 character input" in {
+        val input =
+          Map(identifier -> "XMPPT00000001234")
+        val expectedErrors =
+          Seq(FormError(identifier, lengthError))
+
+        testFailedValidationErrors(input, expectedErrors)
+      }
+
+      "provided with an identifier which doesn't start with XMPTT" in {
```
Comment (paulblakehmrc) [inline test/uk/gov/hmrc/plasticpackagingtax/returns/forms/ClientIdentifierSpec.scala:72]: this test needs tweaking

File: app/uk/gov/hmrc/plasticpackagingtax/returns/controllers/actions/AuthAgentAction.scala [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    block: AuthenticatedRequest[A] => Future[Result]
+  ): Future[Result] = {
+    implicit val hc: HeaderCarrier =
+      HeaderCarrierConverter.fromRequestAndSession(request, request.session)
+
+    val authorisation = authTimer.time()
+
+    authorised(AffinityGroup.Agent)
+      .retrieve(authData) {
+        case credentials ~ name ~ email ~ externalId ~ internalId ~ affinityGroup ~ allEnrolments ~ agentCode ~
```
Comment (paulblakehmrc) [inline app/uk/gov/hmrc/plasticpackagingtax/returns/controllers/actions/AuthAgentAction.scala:63]: Optional - could some of these retrievals be extracted and shared with AuthAction where they're replicated?

**Review by paulblakehmrc — CHANGES_REQUESTED** [review]
LGTM. Main item requiring a change is the PPT regex.

File: app/uk/gov/hmrc/plasticpackagingtax/returns/controllers/actions/AuthAgentAction.scala (hunk 3 above)
Comment (tonymccrae-ee) [reply app/uk/gov/hmrc/plasticpackagingtax/returns/controllers/actions/AuthAgentAction.scala:63]: Almost certainly.
There are elements of merge conflict fear and Scala fear here.
I'd like to get merged, validate this and then look for some clean ups if that's ok.

File: app/uk/gov/hmrc/plasticpackagingtax/returns/forms/agents/ClientIdentifier.scala [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+object ClientIdentifier extends CommonFormValidators {
+
+  implicit val format: OFormat[ClientIdentifier] = Json.format[ClientIdentifier]
+
+  val identifier           = "identifier"
+  val identifierEmptyError = "agents.client.identifier.empty.error"
+  val formatError          = "agents.client.identifier.format.error"
+  val lengthError          = "agents.client.identifier.length.error"
+
+  private val validFormatPattern = Pattern.compile("^XMPPT\\d{10}$")
```
[the remark this answers is not in the document]
Comment (tonymccrae-ee) [reply app/uk/gov/hmrc/plasticpackagingtax/returns/forms/agents/ClientIdentifier.scala:37]: Intetesting - all or examples are XM - will mix those up abit.

File: test/uk/gov/hmrc/plasticpackagingtax/returns/forms/ClientIdentifierSpec.scala (hunk 2 above)
Comment (tonymccrae-ee) [reply test/uk/gov/hmrc/plasticpackagingtax/returns/forms/ClientIdentifierSpec.scala:72]: rename to with correct prefix.

File: app/uk/gov/hmrc/plasticpackagingtax/returns/controllers/actions/AuthAgentAction.scala (hunk 3 above)
Comment (paulblakehmrc) [reply app/uk/gov/hmrc/plasticpackagingtax/returns/controllers/actions/AuthAgentAction.scala:63]: yes, that's fine.

**Review by paulblakehmrc — APPROVED** [review]
